### PR TITLE
fix #222 scrollbar covered cell content

### LINKED ISSUE #222: scrollbar covered cell content
When columns width === grid width and there's a vertical scrollbar, the scrollbar will cover the last cell of every row.
Similarly, when rows Height === grid height and theres a horizontal scrollbar, it will cover the last cell of every column.

here's an example: http://plnkr.co/edit/lV9gaaS34xoybhfiWipJ

a
b

Seems the problem was from Grid.js  [here](https://github.com/bvaughn/react-virtualized/blob/master/source/Grid/Grid.js#L436) and [here](https://github.com/bvaughn/react-virtualized/blob/master/source/Grid/Grid.js#L440)

May be change it to something like this can solve the problem?

```js
    if (totalColumnsWidth + (totalRowsHeight > height ? this._scrollbarSize : 0) <= width) {
      gridStyle.overflowX = 'hidden'
    }

    if (totalRowsHeight + (totalColumnsWidth > width ? this._scrollbarSize : 0) <= height) {
      gridStyle.overflowY = 'hidden'
    }
```

### PR DESCRIPTION
**Pull request by anjianshi**

### CODE REVIEW

File: source/Grid/Grid.js [hunk 1]
```diff
@@ -204,6 +204,8 @@ export default class Grid extends Component {
 
     // See defaultCellRangeRenderer() for more information on the usage of this cache
     this._cellCache = {}
+
+    this._scrollbarSize = getScrollbarSize()
```
Comment (bvaughn) [inline source/Grid/Grid.js]: I don't believe this is a good idea. I think it could cause problems with server-side rendering.

Comment (bvaughn) [inline source/Grid/Grid.js]: Elaborating a bit- I think the `inDom` check for `dom-helpers` will prevent scrollbar size from ever being set in that case. Because at the time the component is _created_ it won't be in the DOM.

Hmm... maybe we could address this concern by:
* Adding a fallback to `render` to handle a potentially undefined size (eg `const scrollbarSize = this._scrollbarSize || 0`)
* Adding a check in `componentDidMount` for an undefined `this._scrollbarSize` and re-initializing if necessary

**Comment by bvaughn** [comment]
Friendly bump @anjianshi. If you're willing to make the suggested PR changes for this~ great! If not let me know and I'll close the PR and tackle the issue when I get some time. :)

**Comment by anjianshi** [comment]
Oh, sorry! these days I was very busy, tomorrow I will try to improve the PR.

**Comment by bvaughn** [comment]
No problem! I understand being busy. :)

File: source/Grid/Grid.js (hunk 1 above)
Comment (anjianshi) [inline source/Grid/Grid.js]: @bvaughn  I think this is not good. Because the first render will not be correct. May be we can skip set the overflow style in first render? I think this will not affect #116.

File: source/Grid/Grid.test.js [hunk 2]
```diff
@@ -120,43 +120,49 @@ describe('Grid', () => {
   describe('shows and hides scrollbars based on rendered content', () => {
     const scrollbarSize = getScrollbarSize()
 
+    function renderTwice (markup) {
```
Comment (bvaughn) [inline source/Grid/Grid.test.js]: Why is this helper necessary? (Why is rendering twice required by these tests?)

Comment (anjianshi) [inline source/Grid/Grid.test.js]: Because in first render, `overflowX` or `overflowY` always won't be set
(because `this._scrollbarSize` was `undefined`)

Comment (anjianshi) [inline source/Grid/Grid.test.js]: Now in first render, overflowX and overflowY won't be set, so it can produce right result, whatever server-side rending or client-side rending.
And in follow up renders, because `this._scrollbarSize` was set, `overflowX` and `overflowY` will be set normally, to prevent #116 's problem

Comment (bvaughn) [inline source/Grid/Grid.test.js]: This is why I previous suggested what I did- initializing it in `componentWillMount` and then...
> * Adding a check in `componentDidMount` for an undefined `this._scrollbarSize` and re-initializing if necessary

This would handle the server-side rendering concern.

We should not require external code (test code or production code) to call `forceUpdate` in order for the correct layout behavior. `Grid` should update itself if it detects that condition.

Comment (anjianshi) [inline source/Grid/Grid.test.js]: but without this helper, the test won't pass.

Comment (bvaughn) [inline source/Grid/Grid.test.js]: Re-read my suggestion above? :)

Comment (anjianshi) [inline source/Grid/Grid.test.js]: In first render, overflowX or Y always won't be set, so there's noting to test.
The purpose of call forceUpdate() is testing the second render's result.
Or there's other way to do this?

Or we should `adding a fallback to render to handle a potentially undefined size (eg const scrollbarSize = this._scrollbarSize || 0)` as you say?
But this will cause the first render got wrong result about overflowX / overflowY.

sorry, my english is not very well. :)

Comment (bvaughn) [inline source/Grid/Grid.test.js]: It's not acceptable for react-virtualized users to have to call `forceUpdate` to get correct layout behavior. The server-side rendering edge-case we're talking about can be detected and handled by `Grid` automatically by checking for `this._scrollbarSize == null`. At that point your tests- (and more important, users using `Grid`- won't have to call `forceUpdate` themselves.

Does this make better sense?

Comment (bvaughn) [inline source/Grid/Grid.test.js]: What I originally suggested was to...
* Measure scrollbar size in [`componentWillMount`](https://facebook.github.io/react/docs/component-specs.html#mounting-componentwillmount) so that client-side content will only be measured once, right before rendering the first time.
* Check for an undefined scroll bar size in [`componentDidMount`](https://facebook.github.io/react/docs/component-specs.html#mounting-componentdidmount) so that server-side rendered content will be re-measured for scrollbar size at the first opportunity. In this second case, if scrollbar isn't measured until `componentDidMount`, then the `Grid` should `forceUpdate` (or `setState`) itself to ensure that it renders one more time with the adjusted scrollbar size.

Comment (bvaughn) [inline source/Grid/Grid.test.js]: I have a little time today while flying to just pick this up. I'll base it on what you've done and make the changes I've suggested. I'll give you credit for the idea/suggestion in release notes. :)

Comment (anjianshi) [inline source/Grid/Grid.test.js]: Understand. Happy to see these improve, and thanks for your excellent work!
